#98 ダイスクラスの作成

t-nagatsuka-gr opened the pull request.

t-nagatsuka-gr committed:
#98 ダイスクラスの作成

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,32 @@
+    private const int DicePipValue6 = 6;
+
+    private int[] numbers;
+    private Random random;
```
基本的に全てのメンバ変数にコメントを記載してください。

t-nagatsuka-gr committed:
#98 ダイスクラスの作成
メンバ変数のコメント追加

t-nagatsuka-gr committed:
#98 ダイスクラスの作成
定数名の変更
メンバ変数名の変更
乱数保持のための配列を定義

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,47 @@
+
+public class Dice
+{
+    // サイコロの出目に出力する値   
```
こういった値の列挙はenum型を使用するとスマートです。
使い方が少し難しいので、次回説明だけ行います。

a-kiyo-g-relations replied:
enum使って頂きましたが、一応説明は追加で行います。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,47 @@
+        DicePip4, DicePip5, DicePip6 };
+
+    //サイコロの乱数を保持するための配列
+    private int[] diceNumbers = new int[3];
```
こちらはサイコロを振って出目の保持用でしょうか。
そうであれば、このクラスは「サイコロ1つ」を表すはずなので、配列ではなく通常の整数型かと思います。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,47 @@
+    private const int DicePip6 = 6;
+
+    // サイコロの出目の定数を入れるための配列
+    private int[] pipDice = new int[] { 
```
この変数は定数で良いのではないでしょうか。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,47 @@
+        DicePip1, DicePip2, DicePip3, 
+        DicePip4, DicePip5, DicePip6 };
+
+    //サイコロの乱数を保持するための配列
```
//後のスペースが抜けています。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,47 @@
+    // 乱数を指定の数ループし格納
+    public int[] GenerateRandomNumbers()
+    {
+        int count = 3; // 生成する乱数の数
```
上でも書きましたが、このクラスは「サイコロ1つ」を表すため、配列にはならないかと思います。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    // 乱数を生成する
+    public int GetRandomDiceNumber()
```
「サイコロを振る(ランダムな値を格納する)」
「前回の出目を取得する」
というメソッド2つに分割してください。

一緒にしてしまうと、出目の値をクラス外部で保存しなくてはいけません。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,47 @@
+        }
+        return diceNumbers;
+    }
+}
```
細かいですが、ファイル末尾は}で終わらせるのではなく、1行空行を入れた方がベターです。
理由は別途説明します。

t-nagatsuka-gr committed:
#98 ダイスクラスの作成
修正点
・前回の出目を保持する変数追加
・定数定義変更
・「サイコロを振る(ランダムな値を格納する)」
  「前回の出目を取得する」メソッド2つに分割
・ファイル末尾二行改行

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,56 @@
+    // サイコロの出目の定数を入れるための配列
+    private PipDice[] pipDice = new PipDice[] {
+        PipDice.DicePipOne, PipDice.DicePipTwo, PipDice.DicePipThree,
+        PipDice.DicePipFour, PipDice.DicePipFive, PipDice.DicePipSix };
```
enumにしたらこの配列は不要になるかと思います。
enumの値を配列化するメソッドが存在するため、それを使えば良いです。
(ランダム値を取得するメソッド内で配列化すれば良いので、配列として定義しておく必要はない。むしろ二重定義になってしまうため、今後の拡張で邪魔になりうる)

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,56 @@
+    // 前回の出目を取得する
+    public int GetRollValue()
+    {
+            return rollValue;   
```
インデントずれ

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,56 @@
+    // 直前の出目を保持する変数
+    private PipDice rollNumber;
+    // 前回の出目を保持する変数
+    private int rollValue;
```
保持する変数はrollNumberかrollValueのどちらかで良いかと思います。
考え方としては、enum型で定義しておいて必要になったらint型に変換する。もしくはint型で数字だけ保持して必要になったらenum型に変換する。という2つが考えられます。
(いつでも任意で変換が可能なので、両方の値を保持する必要はない)

今回に関しては、int型で保持すれば良いかと思います。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,56 @@
+        DicePipFour = 4,
+        DicePipFive = 5,
+        DicePipSix = 6
+    }
```
enum型で定義した場合、アクセスは常に `PipDice.DicePipOne` または `Dice.PipDice.DicePipOne` のようになるため、定義名にかなり重複が発生します。
```c#
public class Dice // ダイスクラスの
{
    public enum Pip // 出目の
    {
        One = 1 // 1の目
    }
}

// アクセス時
Dice.Pip.One // ダイスの出目の1だとわかる
``` 
みたいに、親の名前空間を意識した定義名にするとスマートです。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    // サイコロを振り、ランダムな値を返し、RollNumberに保持
+    public int Roll()
```
挙動がぼやけるので、戻り値なしのメソッドで良いと思います。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,56 @@
+        random = new Random();
+    }
+
+    // サイコロを振り、ランダムな値を返し、RollNumberに保持
```
コメントは「(変数名)に保持」より「(どんな値かの説明)を保持」とした方がより良いです。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    // 乱数を生成して返す
+    public PipDice GetRandomNumber()
```
このメソッドはpublicではない方が良いかと思います。
基本的には内部でしか使用しないはずです。

t-nagatsuka-gr committed:
#98 diceクラスの作成
・配列定義をメソッド内で実行に修正
・コメントの修正
・修飾子の修正
・enum型の定義名を修正

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,49 @@
+    // 前回の出目を取得する
+    public int GetRollValue()
+    {
+        Roll();
```
これだと呼ぶたびに値が変わってしまいます。
Rollは不要かと思います。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,49 @@
+    public void Roll()
+    {
+        Pip rollNumber = GetRandomNumber(); // ランダム値を保持
+        rollValue = (int)rollNumber; // 列挙型を整数にキャストして代入
```
ここのコメント「ランダム値を保持」「列挙型を整数にキャストして代入」はコードから自明なのでなくても良いと思います。
また、特に理由のない場合は行末コメントより1行使ってコメントを記載した方が良いです。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,49 @@
+    {
+        Array diceValues = Enum.GetValues(typeof(Pip)); // 列挙型の値を取得
+        int randomIndex = random.Next(diceValues.Length);
+        return (Pip)diceValues.GetValue(randomIndex); // 乱数を Pip 列挙型に変換して返す
```
enum(Pip)型に変換する必要はありますか？
そのままint型でも良い気がします。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -0,0 +1,49 @@
+        rollValue = (int)rollNumber; // 列挙型を整数にキャストして代入
+    }
+
+    // 乱数を生成して返す
```
厳密には「乱数を生成する」ではなく「サイコロの出目をランダムに取得する」かと思います。

ここは好みですが、Rollの中に処理をまとめても良いかとは思います。
(今回はバラして良いと思います)

t-nagatsuka-gr committed:
#98 ダイスクラスの作成
・コメントの見直し
・Rollメソッド名をRollDiceに修正
